[os_log][SIL Optimization] Make new os log APIs @_transparent and change OSLogOptimization pass so that inlining is not performed.

ravikandhadai opened the pull request:
@atrick  @eeckstein, I have made a change to the OSLogOptimization pass to address an issue raised by @gottesmm. Basically, I have removed inlining from the pass, which caused problems as the pass was in the mandatory pipeline but wasn't using Mandatory inlining (which happens earlier). The reason for that was that the pass used  @_semantics attribute to identify log calls, extract the custom string interpolation argument  of the call, and identify the instructions that must be constant evaluated. The call and the @_semantics attribute will be lost by mandatory inlining.

This commit changes how the instructions to constant evaluate are identified. The new approach uses the "initializer" calls of the custom string interpolation type: 'OSLogMessage' to drive the constant evaluation and folding. The initializer calls need not be inlined and are constant evaluable.   The main benefit of the current approach is that, it will kick in at every place where the type 'OSLogMessage' is instantiated using a string interpolation, regardless of whether or not it is passed to a os_log call.  E.g. the optimization will also apply to cases like the following.

```
    let x: OSLogMessage = "Hello \(1)"
    let y = x.formatString // This will be constant folded. 
```

A note: the diagnostics in this pass will be subsumed and enhanced by a new pass that will enforce the library implementation model (by checking if the implementation is constant evaluable etc.) and also the user model for the log calls (by checking if the log calls are passed a string interpolation).  The diagnostics in this pass will eventually become redundant and should just be asserts/noops.

ravikandhadai committed:
[os_log][SIL Optimization] Make new os log APIs @_transparent and
change OSLogOptimization pass so that inlining is not performed.

ravikandhadai committed:
[os_log][SIL Optimization] Modify the OSLogOptimization pass to run
on ownership SIL.

ravikandhadai commented:
@swift-ci Please test

gottesmm commented on lib/SILOptimizer/Mandatory/OSLogOptimization.cpp:
```diff
@@ -264,28 +268,38 @@ evaluateOrSkip(ConstExprStepEvaluator &stepEval,
     SILInstruction *currInst = &(*currI);
 
+    // Break if currInst is a release_value of 'oslogMessage'.
+    if (auto *releaseValueInst = dyn_cast<ReleaseValueInst>(currInst)) {
```
1. For this to work when ownership is enabled, you need to at least handle destroy_value and probably need to look through copy_value.
2. Is this a struct or a class? If it is a class, you probably also need strong_release.

ravikandhadai replied:
> So to be clear: you marked the initializer with a semantic call so you could just constant fold it instead of inlining it?

Yes, that's right. One thing I would add is that the initializer was always constant evaluated but was mandatorily inlined earlier and then evaluated. Now it is marked with @_semantics and not mandatorily inlined (it can still be constant evaluated), and more importantly it is used to identify the range of instructions to evaluate.

ravikandhadai replied:
> * For this to work when ownership is enabled, you need to at least handle destroy_value and probably need to look through copy_value.
> * Is this a struct or a class? If it is a class, you probably also need strong_release.

This is a struct. I see. Thanks for the update, I will watch out for that.  Actually, this condition is more of a safe guard and in reality  the evaluation should break even before this instruction  as there will instructions that it cannot constant evaluate (due to the implementation of os_log) and will bail out earlier.

gottesmm reviewed (commented):
So to be clear: you marked the initializer with a semantic call so you could just constant fold it instead of inlining it?

ravikandhadai commented:
@swift-ci Please test linux platform

gottesmm commented:
Can you add the destroy-value and copy value plz

ravikandhadai commented:
> Can you add the destroy-value and copy value plz

Okay. Will `copy_value` happen just before destroy_value? In that case, it and also destroy_value can be used as a break condition

ravikandhadai commented:
I pushed a commit that makes the OSLogOptimization pass work on ownership SIL. The following are the two main changes: 
 * Fold results of  instructions like `destructure_tuple` that could return multiple results and some of which are foldable.
 * Include `destroy_value` and `copy_value` in end of lifetime checks that is used to stop evaluation. 

There is still a problem in this pass that causes ownership verification to fail. (Basically, the string initialization instructions that created to fold string literals are not released.)  Excepting this, the pass works on ownership SIL and when stripping happens after serialization.

ravikandhadai commented:
@swift-ci Please test

gottesmm commented:
@ravikandhadai was the lack of releasing of the initialization instructions a /real/ bug?

gottesmm commented:
Also, you may want to look at destructure_struct as well.

ravikandhadai commented:
> @ravikandhadai was the lack of releasing of the initialization instructions a /real/ bug?

Yes, it was a leak, but only discovered by the ownership verifier. Basically, when an instruction like `%str = struct_extract S.str` is folded and replaced by a string constant, the optimizer should also add a release/destroy _value instruction. I have updated the code so that cases like these are handled. Now the optimization pass also works with ownership SIL (but for <rdar://51512044>).

The main changes for this is in the function `destroyFoldedValueAtEndOfUse` which is called by `substituteConstants`.

ravikandhadai commented:
@swift-ci Please test

ravikandhadai commented:
@swift-ci Please test macOS Platform

ravikandhadai commented:
@swift-ci Please smoke test macOS Platform

ravikandhadai commented:
@swift-ci Please smoke test macOS Platform

ravikandhadai commented:
@swift-ci Please test macOS Platform

ravikandhadai commented:
@swift-ci Please clean test macOS Platform
